### DISCUSSION TITLE
Clearing data on HTTP 404

### DISCUSSION THREAD

**Discussion by joshkel** [vercel/swr#998] [Q&A] [answered]
If useSWR gets an error after successfully getting data, it keeps the previously cached (stale) data while also reporting the error. This is normally okay, except for one scenario: If a useSWR request results in an HTTP 404 response, I want to treat it as an error, but I also want it to recognize the fact that the data no longer exists and remove it from the cache.

Is there a clean / recommended way of doing this? The best approach I've come up with so far is to hack together an `onError` and `onErrorRetry` handler, but it feels pretty ugly.

```
function onError(err: Error, key: string) {
  if (err instanceof BackendError && err.response.statusCode === StatusCodes.NOT_FOUND) {
    // Use mutate to clear the local data. First check the cache to avoid an infinite
    // loop. mutate clears the error, so pass shouldRevalidate true so that useSWR
    // will retry and see a new error.
    if (cache.get(key)) {
      mutate(key, null, true).catch(promiseCatchAll);
    }
  }
}

function onErrorRetry<T>(
  err: Error,
  key: string,
  config: ConfigInterface<T, Error>,
  revalidate: revalidateType,
  opts: RevalidateOptionInterface
) {
  if (err instanceof BackendError && err.response.statusCode === StatusCodes.NOT_FOUND) {
    // If we know that the item isn't there, there's no reason to retry.
    return;
  }

  // Continue with standard isDocumentVisible(), retryCount, exponential backoff...
}

useSWR(key, fetcher, { onError, onErrorRetry });
```

**Comment by ErnestBrandi** [comment, marked as answer]
Have you seen this page of the doc ? (https://swr.vercel.app/docs/error-handling)
Seems like it addresses your issue pretty well.

**Reply by joshkel** [reply]
I saw that, but I don't think it answers my question. My root problem seems to be that a request in SWR can return data _or_ report an error, and if SWR issues a request that reports an error, then it continues to return the last valid (stale) data. But, for a 404 in particular, I want to report an error (because it is an error code) _and_ clear any previous stale data (because the 404 indicates that we now know that there's no valid data).

**Reply by ErnestBrandi** [reply]
Why not returning a custom value on error in the fetcher function ?
```
const fetcher = async url => {
  const res = await fetch(url)
  if (res.status === 404) {
    return null;
  }
  return res.json()
}
```
Also you could mutate data locally on error : `if(error) mutate('api/call', null, false)`

**Reply by joshkel** [reply]
I wondered about the custom value on 404 error. That loses the ability to view the 404 error itself. And it means changing anything that calls `useSWR` - instead of "we have an error, or valid data, or `undefined` means we're still loading" it's now "we have an error, or valid data, or `undefined` means it's still loading, or `null` means no error but no data."

Maybe that is the best way to model things, though.

Thanks.

**Reply by ErnestBrandi** [reply]
You can use this special fetcher for that one SWR call only. Also, the second solution do not hide the 404 error.

**Reply by joshkel** [reply]
Thanks.

When I tested the second solution, calling `mutate` with revalidate `false` does clear the error.

**Comment by michaelfortunato** [comment]
The docs should be more explicit that the cache does *not* clear out the last successful response if the subsequent response is an error. I'd like to see that in the docs on mutate. Happy for this thread as I was completely confused as to why my cache response was stale.

**Reply by michaelfortunato** [reply]
Do you all think it would make sense to file an issue on SWR to update the docs?

**Reply by shuding** [reply]
This is rather an expected behavior, but the docs should be more explicit, feel free to file an issue!

**Reply by shuding** [reply]
The reason that SWR wouldn’t clear the stale data when there’s an error is, after a failed refetch many apps show the stale data and the error at the same time. And usually that gives a better UX for things like polling.

**Reply by michaelfortunato** [reply]
Thanks! Ok I might file an issue.
